### PR TITLE
planner, table: Disallow update self (natural) join on partitioning columns (#31629)

### LINKED ISSUE #31629: TiDB Update with Natural join on Partitioned table results are not consistent with mysql.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists t1;

CREATE TABLE t1(partkey int,nokey int
) PARTITION BY HASH(partkey) PARTITIONS 3;

INSERT INTO t1 VALUES (1, 1) , (10, 10);

SELECT * FROM t1;
UPDATE t1 AS A NATURAL JOIN t1 B SET A.partkey = 2, B.nokey = 3;
SELECT * FROM t1;
```

### 2. What did you expect to see? (Required)
```sql
mysql> SELECT * FROM t1;
+---------+-------+
| partkey | nokey |
+---------+-------+
|       1 |     1 |
|      10 |    10 |
+---------+-------+
2 rows in set (0.00 sec)
mysql> UPDATE t1 AS A NATURAL JOIN t1 B SET A.partkey = 2, B.nokey = 3;
ERROR 1706 (HY000): Primary key/partition key update is not allowed since the table is updated both as 'A' and 'B'
```
### 3. What did you see instead (Required)
```sql
mysql> SELECT * FROM t1;
+---------+-------+
| partkey | nokey |
+---------+-------+
|       1 |     1 |
|      10 |    10 |
+---------+-------+
2 rows in set (0.00 sec)

mysql> UPDATE t1 AS A NATURAL JOIN t1 B SET A.partkey = 2, B.nokey = 3;
Query OK, 4 rows affected (0.00 sec)
Rows matched: 4  Changed: 4  Warnings: 0

mysql> SELECT * FROM t1;
+---------+-------+
| partkey | nokey |
+---------+-------+
|       2 |     3 |
|       2 |     3 |
|       2 |     1 |
|       2 |    10 |
+---------+-------+
4 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)
```sql
| Release Version: v5.4.0-alpha-264-g6efa36df6
Edition: Community
Git Commit Hash: 6efa36df6cff325106f67ecfe3d79816ba37ca6a
Git Branch: master
UTC Build Time: 2021-12-28 02:03:55
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #31629

Problem Summary:
Self natural join was allowed on partitioning columns, resulting in UPDATE could even duplicate rows.

### What is changed and how it works?
Dissallow update for natural joins containing partitioning columns

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Disallowing update on self natural join on partitioning columns.
```

### CODE REVIEW

File: table/tables/partition.go [hunk 1]
```diff
@@ -903,6 +903,29 @@ func (t *partitionedTable) PartitionExpr() (*PartitionExpr, error) {
 	return t.partitionExpr, nil
 }
 
+func (t *partitionedTable) GetPartitionColumnNames() []model.CIStr {
```
Comment (djshow832) [inline table/tables/partition.go:906]: Can you add a unit test for this function with various partition types?

Comment (mjonss) [reply table/tables/partition.go:906]: Done! Thank you for the suggestion.

**Comment by mjonss** [comment]
Please also test something like:
`UPDATE t2 SET a='aaa' where a in (SELECT a from t2);` which should not be allowed.

**Review by djshow832 — APPROVED** [review]

**Comment by mjonss** [comment]
> Please also test something like: `UPDATE t2 SET a='aaa' where a in (SELECT a from t2);` which should not be allowed.

That will be reported as a separate bug, since it seems to be allowed also for non-partitioned tables.

**Review by tiancaiamao — APPROVED** [review]

**Comment by qw4990** [comment]
> /cc @qw4990 do you have time to review this?

I reviewed these file changes, it's LGTM.

**Pull request merged**